src/groth16/verifier/batch.rs

```diff
@@ -167,4 +173,117 @@ where
+        }
+
+        struct Accumulator<E: MultiMillerLoop> {
+            gammas: Vec<E::Fr>,
+            delta: E::G1,
+            y: E::Fr,
+            ml_result: Option<E::Result>,
+        }
+
+        impl<E: MultiMillerLoop> Accumulator<E> {
+            fn new(ic_len: usize) -> Self {
+                Accumulator {
+                    gammas: vec![E::Fr::zero(); ic_len],
+                    delta: E::G1::identity(),
+                    y: E::Fr::zero(),
+                    ml_result: None,
+                }
+            }
+        }
+
+        let ic_len = vk.ic.len();
+
+        let acc = self
+            .items
+            // This chunk size was obtained heuristically.
+            .par_chunks(8)
+            .map(|items| {
+                let mut acc = Accumulator::<E>::new(ic_len);
+                let mut ml_terms: Vec<(E::G1Affine, E::G2Prepared)> = vec![];
+                let z = loop {
+                    let z = E::Fr::random(&mut OsRng);
+                    if !z.is_zero_vartime() {
+                        break z;
+                    }
+                };
+                let mut cur_z = z;
+                for Item { proof, inputs } in items {
+                    acc.gammas[0] += &cur_z;
+                    for (a_i, acc_gamma_i) in
+                        Iterator::zip(inputs.iter(), acc.gammas.iter_mut().skip(1))
+                    {
+                        *acc_gamma_i += &(cur_z * a_i);
+                    }
+                    acc.delta += proof.c * cur_z;
+                    acc.y += &cur_z;
+                    ml_terms.push(((proof.a * cur_z).into(), (-proof.b).into()));
+
+                    cur_z *= z;
+                }
+                let ml_terms = ml_terms.iter().map(|(a, b)| (a, b)).collect::<Vec<_>>();
+                acc.ml_result = Some(E::multi_miller_loop(&ml_terms[..]));
+                acc
+            })
+            .reduce(
+                || Accumulator::<E>::new(ic_len),
+                |mut a, b| {
+                    for (a, b) in a.gammas.iter_mut().zip(b.gammas.into_iter()) {
+                        *a += b;
+                    }
+                    a.delta += b.delta;
+                    a.y += b.y;
+                    a.ml_result = match (a.ml_result, b.ml_result) {
+                        (Some(a), Some(b)) => Some(a + b),
+                        (Some(a), None) | (None, Some(a)) => Some(a),
+                        (None, None) => None,
+                    };
+                    a
+                },
+            );
+
+        // TODO: could use a multiexp (Bos-Coster maybe?)
+        let psi = vk
+            .ic
+            .iter()
+            .zip(acc.gammas.into_iter())
+            .map(|(&psi_i, acc_gamma_i)| psi_i * acc_gamma_i)
+            .sum();
+
+        match acc.ml_result {
+            None => Ok(()),
+            Some(mut ml_result) => {
+                ml_result += E::multi_miller_loop(&[
```

Comment:
`acc.ml_result` will be `None` when there are no items in the batch. In this case, we can avoid the computation of `psi` to make an empty batch effectively free:
```suggestion
        match acc.ml_result {
            None => Ok(()),
            Some(mut ml_result) => {            
                // TODO: could use a multiexp (Bos-Coster maybe?)
                let psi = vk
                    .ic
                    .iter()
                    .zip(acc.gammas.into_iter())
                    .map(|(&psi_i, acc_gamma_i)| psi_i * acc_gamma_i)
                    .sum();

                ml_result += E::multi_miller_loop(&[
```